test-framework/sudo-compliance-tests/src/sudoedit/limits.rs

```diff
@@ -0,0 +1,162 @@
+                );
+            } else {
+                // I don't know why ogsudo gives this error -- probably because opening failed
+                assert_contains!(output.stderr(), "No such file or directory");
+            }
+        } else {
+            assert_contains!(
+                output.stderr(),
+                "cannot open a file in a path writable by the user"
+            );
+        }
+        output.assert_exit_code(1);
+    }
+}
+
+#[test]
+fn can_edit_writable_paths_as_root() {
+    // note: we already have tests that sudoedit "works" so we are skipping
+    // the content check here---the point here is that sudoedit does not stop
+    // the user.
+
+    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
+        .user(USERNAME)
+        .directory(Directory("/tmp/bar").chmod("755"))
+        .file(DEFAULT_EDITOR, TextFile(EDITOR_DUMMY).chmod(CHMOD_EXEC))
+        .build();
+
+    let file = "/tmp/foo.sh";
+    Command::new("sudoedit")
+        .arg(file)
+        .output(&env)
+        .assert_success();
+}
+
+#[test]
+fn cannot_edit_symlinks() {
+    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
+        .user(USERNAME)
+        .file(DEFAULT_EDITOR, TextFile(EDITOR_DUMMY).chmod(CHMOD_EXEC))
+        .build();
+
+    let file = "/usr/bin/sudoedit";
+
+    let output = Command::new("sudoedit")
+        .as_user(USERNAME)
+        .arg(file)
+        .output(&env);
+
+    assert_contains!(output.stderr(), "editing symbolic links is not permitted");
+
+    output.assert_exit_code(1);
+}
+
+#[test]
+fn cannot_edit_files_target_user_cannot_access() {
+    let file = "/test.txt";
+
+    let env = Env(SUDOERS_ALL_ALL_NOPASSWD)
+        .user(USERNAME)
+        .user(OTHER_USERNAME)
+        .group(USERNAME)
+        .group(OTHER_USERNAME)
+        .file(DEFAULT_EDITOR, TextFile(EDITOR_DUMMY).chmod(CHMOD_EXEC))
+        .file(
+            file,
+            TextFile("")
+                .chown(format!("{USERNAME}:{ROOT_GROUP}"))
+                .chmod("460"),
+        )
+        .build();
+
+    let test_cases = [
+        // incorrect user
+        &["-u", OTHER_USERNAME][..],
+        // correct user, but does not have write permits
+        &["-u", USERNAME][..],
+        // incorrect group
+        &["-u", OTHER_USERNAME, "-g", USERNAME][..],
+        // group permission doesn't override matching user permissions
+        &["-u", USERNAME, "-g", ROOT_GROUP][..],
+        &["-g", ROOT_GROUP][..],
+    ];
+
+    for args in test_cases {
+        let output = Command::new("sudoedit")
+            .args(args)
+            .arg(file)
+            .as_user(USERNAME)
+            .output(&env);
+
+        assert_contains!(output.stderr(), "Permission denied");
+        output.assert_exit_code(1);
+    }
+}
+
+#[test]
+fn can_edit_files_target_user_or_group_can_access() {
+    // note: we already have tests that sudoedit "works" so we are skipping
+    // the content check here---the point here is that sudoedit does not stop
```

Comment:
What's with these `---` without spaces around them?

Reply:
A habit from LaTeX days, where `---` is an emdash; and then the Chicago Manual of Style calls for no space around them. I promise to you I did not sure AI to write these comments.